File: day-02/main.py
import os

def score_game(input_lines):
    outcome_matrix = [[3,6,0],[0,3,6],[6,0,3]]
    move_matrix = [1,2,3]
    total_score = 0
    for game in input_lines:
        them, us = game.split(" ")
        if them == "A":
            if us == 'X':
                total_score += outcome_matrix[0][0] + move_matrix[0]
            if us == 'Y':
                total_score += outcome_matrix[0][1] + move_matrix[1]
            if us == 'Z':
                total_score += outcome_matrix[0][2] + move_matrix[2]
        if them == "B":
            if us == 'X':
                total_score += outcome_matrix[1][0] + move_matrix[0]
            if us == 'Y':
                total_score += outcome_matrix[1][1] + move_matrix[1]
            if us == 'Z':
                total_score += outcome_matrix[1][2] + move_matrix[2]
        if them == "C":
            if us == 'X':
                total_score += outcome_matrix[2][0] + move_matrix[0]
            if us == 'Y':
                total_score += outcome_matrix[2][1] + move_matrix[1]
            if us == 'Z':
                total_score += outcome_matrix[2][2] + move_matrix[2]
    return total_score


def new_score_game(input_lines):
    move_matrix = [[3,1,2],[1,2,3],[2,3,1]]
    outcome_matrix = [0,3,6]
    total_score = 0
    for game in input_lines:
        them, us = game.split(" ")
        if them == "A":
            if us == 'X':
                total_score += move_matrix[0][0] + outcome_matrix[0]
            if us == 'Y':
                total_score += move_matrix[0][1] + outcome_matrix[1]
            if us == 'Z':
                total_score += move_matrix[0][2] + outcome_matrix[2]
        if them == "B":
            if us == 'X':
                total_score += move_matrix[1][0] + outcome_matrix[0]
            if us == 'Y':
                total_score += move_matrix[1][1] + outcome_matrix[1]
            if us == 'Z':
                total_score += move_matrix[1][2] + outcome_matrix[2]
        if them == "C":
            if us == 'X':
                total_score += move_matrix[2][0] + outcome_matrix[0]
            if us == 'Y':
                total_score += move_matrix[2][1] + outcome_matrix[1]
            if us == 'Z':
                total_score += move_matrix[2][2] + outcome_matrix[2]
    return total_score

if __name__ == '__main__':
    with open('input.txt') as in_fh:
        file_lines = [line.strip() for line in in_fh.readlines()]

    score = score_game(file_lines)
    print(f"Score: {score}")
    new_score = new_score_game(file_lines)
    print(f"New Score: {new_score}")

File: day-02/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func ScoreShape(shape string) int {
	if shape == "X" {
		return 1
	}
	if shape == "Y" {
		return 2
	}
	if shape == "Z" {
		return 3
	}
	panic("didn't recognize the shape")
}

func ScoreOutcome(shapes []string) int {
	them := shapes[0]
	me := shapes[1]
	if them == "A" {
		if me == "X" {
			return 3
		}
		if me == "Y" {
			return 6
		}
		if me == "Z" {
			return 0
		}
	}
	if them == "B" {
		if me == "X" {
			return 0
		}
		if me == "Y" {
			return 3
		}
		if me == "Z" {
			return 6
		}
	}
	if them == "C" {
		if me == "X" {
			return 6
		}
		if me == "Y" {
			return 0
		}
		if me == "Z" {
			return 3
		}
	}
	panic("couldn't recognize one shape or another")
}

func NewScoreOutcome(outcome string) int {
	if outcome == "X" {
		return 0
	}
	if outcome == "Y" {
		return 3
	}
	if outcome == "Z" {
		return 6
	}
	panic("didn't recognize the outcome string")
}

func NewScoreShape(shapes []string) int {
	them := shapes[0]
	outcome := shapes[1]
	if them == "A" {
		if outcome == "X" {
			return 3
		}
		if outcome == "Y" {
			return 1
		}
		if outcome == "Z" {
			return 2
		}
	}
	if them == "B" {
		if outcome == "X" {
			return 1
		}
		if outcome == "Y" {
			return 2
		}
		if outcome == "Z" {
			return 3
		}
	}
	if them == "C" {
		if outcome == "X" {
			return 2
		}
		if outcome == "Y" {
			return 3
		}
		if outcome == "Z" {
			return 1
		}
	}
	panic("didn't recognize either the shape or the outcome")
}

func NewScoreRounds(rounds []string) int {
	totalScore := 0
	for _, moves := range rounds {
		moveArr := strings.Split(moves, " ")
		outcomeScore := NewScoreOutcome(moveArr[1])
		shapeScore := NewScoreShape(moveArr)
		totalScore += shapeScore + outcomeScore
	}
	return totalScore
}

func ScoreRounds(rounds []string) int {
	totalScore := 0
	for _, moves := range rounds {
		moveArr := strings.Split(moves, " ")
		shapeScore := ScoreShape(moveArr[1])
		outcomeScore := ScoreOutcome(moveArr)
		totalScore += shapeScore + outcomeScore
	}
	return totalScore
}

func main() {
	readFile, err := os.Open("input.txt")
	if err != nil {
		fmt.Println(err)
	}
	defer readFile.Close()

	fileScanner := bufio.NewScanner(readFile)
	fileScanner.Split(bufio.ScanLines)

	var fileLines []string

	for fileScanner.Scan() {
		fileLines = append(fileLines, fileScanner.Text())
	}

	totalScore := ScoreRounds(fileLines)
	fmt.Printf("The total score is %v\n", totalScore)

	newTotalScore := NewScoreRounds(fileLines)
	fmt.Printf("The new total score is %v\n", newTotalScore)

}
